File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        low, high = 0, len(nums) - 1

        while low < high:
            mid = low + ((high - low) >> 1)

            if nums[low] <= nums[mid]:
                if nums[low] <= target <= nums[mid]: # pivot part
                    high = mid
                else:
                    low = mid + 1
            else:
                if nums[mid] < target <= nums[high]:
                    low = mid + 1
                else:
                    high = mid
        
        if nums[low] == target:
            return low
        
        return -1

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.go
func search(nums []int, target int) int {
    l := 0
    r := len(nums) - 1
    for l <= r {
        m := l + (r - l) >> 1

        if target == nums[m] {
            return m
        }
        
        // sorted side
        if nums[l] <= nums[m] {
            if target >= nums[l] && target <= nums[m] {
                r = m - 1
            } else {
                l = m + 1
            }

            continue
        }
        
        if target >= nums[m] && target <= nums[r] {
            l = m + 1
        } else {
            r = m - 1                
        }
    }

    return -1
}